waggle-dance-core/src/test/java/com/hotels/bdp/waggledance/mapping/service/impl/PollingFederationServiceTest.java

```diff
@@ -0,0 +1,71 @@
+/**
+ * Copyright (C) 2016-2019 Expedia, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hotels.bdp.waggledance.mapping.service.impl;
+
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import java.util.List;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+import com.google.common.collect.Lists;
+
+import com.hotels.bdp.waggledance.api.model.AbstractMetaStore;
+import com.hotels.bdp.waggledance.api.model.MetaStoreStatus;
+import com.hotels.bdp.waggledance.core.federation.service.PopulateStatusFederationService;
+
+@RunWith(MockitoJUnitRunner.class)
+public class PollingFederationServiceTest {
+
+  @Mock
+  private PopulateStatusFederationService populateStatusFederationService;
+
+  @Test
+  public void pollNotifyOnStateChange() throws Exception {
+    PollingFederationService service = new PollingFederationService(populateStatusFederationService);
+
+    AbstractMetaStore primary = AbstractMetaStore.newPrimaryInstance("p", "uri");
+    AbstractMetaStore federate = AbstractMetaStore.newFederatedInstance("f", "uri");
+    primary.setStatus(MetaStoreStatus.AVAILABLE);
+    federate.setStatus(MetaStoreStatus.AVAILABLE);
+
+    List<AbstractMetaStore> metastores = Lists.newArrayList(primary, federate);
+    when(populateStatusFederationService.getAll()).thenReturn(metastores);
+
+    // first time
+    service.poll();
+    verify(populateStatusFederationService, never()).update(primary, primary);
+    verify(populateStatusFederationService, never()).update(federate, federate);
```

Comment:
On the first run, the `previous` map should only contain null values, and therefore the metastore statuses should not be checked. To verify this, `primary` and `federate` could be mocks. Then, instead of modifying their statuses, the behaviour of `primaryMock.getStatus()` and `federateMock.getStatus()` could be changed. Just a thought :)

Reply:
Hmm we are checking the status [on this line](https://github.com/HotelsDotCom/waggle-dance/pull/162/files#diff-8998b00790a07c6ea9c5d45816f75e6fR48). So we should check that it's checked once instead of twice, right?

Reply:
You're right. And actually we don't really need to test this, this wouldn't be testing behaviour, but implementation of that behaviour. For example, the use of `getStatus()` could be extracted to a variable and then it will be used once anyway. The check that no update is made is definitely enough, so forget what I said, apologies.

Reply:
Do you mean something like this?
```java
when(populateStatusFederationService.getAll()).thenReturn(Collections.singletonList(federateMock));
when(federateMock.getStatus()).thenReturn(MetaStoreStatus.AVAILABLE);

// first time
service.poll();
verify(populateStatusFederationService, never()).update(federateMock, federateMock);
verify(federateMock).getStatus();

//second time
when(federateMock.getStatus()).thenReturn(MetaStoreStatus.UNAVAILABLE);
service.poll();
verify(populateStatusFederationService).update(federateMock, federateMock);
verify(federateMock, times(3)).getStatus(); // 1 the first time and 2 the second time
```